src/Misc/layoutbin/run-helper.sh

```diff
@@ -0,0 +1,55 @@
+#!/bin/bash
+
+# Validate not sudo
+user_id=`id -u`
+if [ $user_id -eq 0 -a -z "$RUNNER_ALLOW_RUNASROOT" ]; then
+    echo "Must not run interactively with sudo"
+    exit 1
+fi
+
+# Run
+shopt -s nocasematch
+
+safe_sleep() {
+    if [ ! -x "$(command -v sleep)" ]; then
+        if [ ! -x "$(command -v ping)" ]; then
+            COUNT="0"
+            while [[ $COUNT != 5000 ]]; do
+                echo "SLEEP" > /dev/null
+                COUNT=$[$COUNT+1]
+            done
+        else
+            ping -c 5 127.0.0.1 > /dev/null
+        fi
+    else
+        sleep 5
+    fi
+}
+
+bin/Runner.Listener run $*
+returnCode=$?
+if [[ $returnCode == 0 ]]; then
+    echo "Runner listener exit with 0 return code, stop the service, no retry needed."
+    echo "0 ORIG"
+elif [[ $returnCode == 1 ]]; then
+    echo "Runner listener exit with terminated error, stop the service, no retry needed."
+    echo "1 ORIG"
+elif [[ $returnCode == 2 ]]; then
+    echo "Runner listener exit with retryable error, re-launch runner in 5 seconds."
+    echo "2 ORIG"
+    safe_sleep
+elif [[ $returnCode == 3 ]]; then
+    # Sleep 5 seconds to wait for the runner update process finish
+    echo "Runner listener exit because of updating, re-launch runner in 5 seconds"
+    echo "3 ORIG"
+    safe_sleep
+elif [[ $returnCode == 4 ]]; then
+    # Sleep 5 seconds to wait for the ephemeral runner update process finish
+    echo "Runner listener exit because of updating, re-launch ephemeral runner in 5 seconds"
+    echo "4 ORIG"
+    safe_sleep
+else
+    echo "Exiting with unknown error code: ${returnCode}"
+fi
+
+exit $returnCode
```

Comment:
feels like the contract between the `helper.sh` and `run.sh` is not clear.
from what i feel, the `helper.sh` might only want to return 0 or 1.
check with Eric for this.

Reply:
That would simplify it for sure - 0 for exit, 1 for restart. I felt like `run-helper.sh` giving those returnCodes would conflict with some bash standards (1 for error), but since one's not supposed to run the `helper` in isolation, I suppose it would make sense 👍